File: Sources/CountriesFlagsObjC/source/Countries/UIImage+NetherlandsAntilles.m
//
//  UIImage+NetherlandsAntilles.m
//  CountriesFlags
//

#import "Countries/UIImage+NetherlandsAntilles.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (NetherlandsAntilles)

+ (UIImage *)flagNetherlandsAntillesWithSize:(CGSize)size {
    UIColor* colorWhite = [UIColor whiteColor];
    UIColor* colorBlue = [UIColor colorWithRed:1.0f/255.0f
                                         green:45.0f/255.0f
                                          blue:135.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:220.0f/255.0f
                                        green:23.0f/255.0f
                                         blue:29.0f/255.0f
                                        alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorRed setFill];
    UIRectFill(CGRectMake(size.width / 1280.0f * 500.0f, 0.0f, size.width / 1280.0f * 285.0f, size.height));
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 3.0f, size.width, size.height / 3.0f));
    
    [colorWhite setFill];
    CGFloat yOffset = size.height / 1280.0f * 5.0f;
    CGFloat starRadius = size.height / 854.0f * 38.0f;
    // 1 column
    CGPoint starCenter = CGPointMake(size.width / 1280.0f * 450.0f, (size.height + yOffset) / 2.0f);
    [[UIBezierPath starAtPoint:starCenter radius:starRadius type:FlagStarTypeFivePointed] fill];
    // 2 column
    starCenter = CGPointMake(size.width / 1280.0f * 569.0f, (size.height / 854.0f * 668.0f + yOffset) / 2.0f);
    [[UIBezierPath starAtPoint:starCenter radius:starRadius type:FlagStarTypeFivePointed] fill];
    starCenter = CGPointMake(size.width / 1280.0f * 569.0f, (size.height / 854.0f * 1048.0f + yOffset) / 2.0f);
    [[UIBezierPath starAtPoint:starCenter radius:starRadius type:FlagStarTypeFivePointed] fill];
    // 3 column
    starCenter = CGPointMake(size.width / 1280.0f * 710.5f, (size.height / 854.0f * 668.0f + yOffset) / 2.0f);
    [[UIBezierPath starAtPoint:starCenter radius:starRadius type:FlagStarTypeFivePointed] fill];
    starCenter = CGPointMake(size.width / 1280.0f * 710.5f, (size.height / 854.0f * 1048.0f + yOffset) / 2.0f);
    [[UIBezierPath starAtPoint:starCenter radius:starRadius type:FlagStarTypeFivePointed] fill];
    // 4 column
    starCenter = CGPointMake(size.width / 1280.0f * 829.0f, (size.height + yOffset) / 2.0f);
    [[UIBezierPath starAtPoint:starCenter radius:starRadius type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+NetherlandsAntilles.swift
//
//  UIImage+NetherlandsAntilles.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_the_Netherlands_Antilles "wikipedia"
     
     Generates the flag of the Netherlands Antilles.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the Netherlands Antilles.
     */
    internal static func flagNetherlandsAntilles(with size: CGSize) -> UIImage {
        let colorWhite = UIColor.white
        let colorBlue = UIColor(red: 1.0 / 255.0,
                                green: 45.0 / 255.0,
                                blue: 135.0 / 255.0,
                                alpha: 1.0)
        let colorRed = UIColor(red: 220.0 / 255.0,
                               green: 23.0 / 255.0,
                               blue: 29.0 / 255.0,
                               alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorWhite.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorRed.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: size.width / 1280.0 * 500.0, y: 0.0),
                          size: CGSize(width: size.width / 1280.0 * 285.0, height: size.height)))

        colorBlue.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 3.0),
                          size: CGSize(width: size.width, height: size.height / 3.0)))

        colorWhite.setFill()
        let yOffset = size.height / 1280.0 * 5.0
        let starRadius = size.height / 854.0 * 38.0
        // 1 column
        var starCenter = CGPoint(x: size.width / 1280.0 * 450.0, y: (size.height + yOffset) / 2.0)
        UIBezierPath.star(at: starCenter, radius: starRadius).fill()
        // 2 column
        starCenter = CGPoint(x: size.width / 1280.0 * 569.0, y: (size.height / 854.0 * 668.0 + yOffset) / 2.0)
        UIBezierPath.star(at: starCenter, radius: starRadius).fill()
        starCenter = CGPoint(x: size.width / 1280.0 * 569.0, y: (size.height / 854.0 * 1048.0 + yOffset) / 2.0)
        UIBezierPath.star(at: starCenter, radius: starRadius).fill()
        // 3 column
        starCenter = CGPoint(x: size.width / 1280.0 * 710.5, y: (size.height / 854.0 * 668.0 + yOffset) / 2.0)
        UIBezierPath.star(at: starCenter, radius: starRadius).fill()
        starCenter = CGPoint(x: size.width / 1280.0 * 710.5, y: (size.height / 854.0 * 1048.0 + yOffset) / 2.0)
        UIBezierPath.star(at: starCenter, radius: starRadius).fill()
        // 4 column
        starCenter = CGPoint(x: size.width / 1280.0 * 829.0, y: (size.height + yOffset) / 2.0)
        UIBezierPath.star(at: starCenter, radius: starRadius).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
